templates/common/rules/rules_test.go

```diff
@@ -0,0 +1,164 @@
+	"github.com/abcxyz/pkg/testutil"
+)
+
+func TestValidateRules(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		name  string
+		scope *common.Scope
+		rules []*spec.Rule
+		want  string
+	}{
+		{
+			name: "rules_are_valid",
+			scope: common.NewScope(map[string]string{
+				"my_var": "foo",
+			}),
+			rules: []*spec.Rule{
+				{
+					Rule:    model.String{Val: "size(my_var) < 5"},
+					Message: model.String{Val: "Length must be less than 5"},
+				},
+			},
+			want: "",
+		},
+		{
+			name: "rules_are_invalid",
+			scope: common.NewScope(map[string]string{
+				"my_var": "foobarbaz",
+			}),
+			rules: []*spec.Rule{
+				{
+					Rule:    model.String{Val: "size(my_var) < 5"},
+					Message: model.String{Val: "Length must be less than 5"},
+				},
+			},
+			want: "rules validation failed:\n\nRule:      size(my_var) < 5\nRule msg:  Length must be less than 5\n",
+		},
+	}
+
+	for _, tc := range cases {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			t.Parallel()
+			ctx := context.Background()
+
+			got := ValidateRules(ctx, tc.scope, tc.rules)
+			if diff := testutil.DiffErrString(got, tc.want); diff != "" {
+				t.Errorf("unexpected result (-got, +want):\n%s", diff)
+			}
+		})
+	}
+}
+
+func TestValidateRulesWithMessage(t *testing.T) {
+	t.Parallel()
+	cases := []struct {
+		name     string
+		scope    *common.Scope
+		rules    []*spec.Rule
+		callBack func(*tabwriter.Writer)
+		want     string
+	}{
+		{
+			name: "rules_valid_no_op_callback",
+			scope: common.NewScope(map[string]string{
+				"my_var": "foo",
+			}),
+			rules: []*spec.Rule{
+				{
+					Rule:    model.String{Val: "size(my_var) < 5"},
+					Message: model.String{Val: "Length must be less than 5"},
+				},
+			},
+			callBack: func(tw *tabwriter.Writer) {},
+			want:     "",
+		},
+		{
+			name: "rules_valid_call_back",
+			scope: common.NewScope(map[string]string{
+				"my_var": "foo",
+			}),
+			rules: []*spec.Rule{
+				{
+					Rule:    model.String{Val: "size(my_var) < 5"},
+					Message: model.String{Val: "Length must be less than 5"},
+				},
+			},
+			callBack: func(tw *tabwriter.Writer) {
+				fmt.Fprintln(tw, "this will not be written since rules are valid")
+			},
+			want: "",
+		},
+		{
+			name: "rules_invalid_no_op_callback",
+			scope: common.NewScope(map[string]string{
+				"name": "bar123",
+				"age":  "1054",
+			}),
+			rules: []*spec.Rule{
+				{
+					Rule:    model.String{Val: "name.matches('^[A-Za-z]+$')"},
+					Message: model.String{Val: "name must only contain alphabetic characters"},
+				},
+				{
+					Rule:    model.String{Val: "int(age) < 130"},
+					Message: model.String{Val: "age must be less than 130"},
+				},
+			},
+			callBack: func(tw *tabwriter.Writer) {
+				fmt.Fprintln(tw, "Rule Violation:")
+			},
+			want: "Rule Violation:\n\nRule:      name.matches('^[A-Za-z]+$')\nRule msg:  name must only contain alphabetic characters\nRule Violation:\n\nRule:      int(age) < 130\nRule msg:  age must be less than 130\n",
```

Comment:
Optional: you can use backticks to create a raw multiline string, so you wouldn't need `\n`.

Reply:
done, that makes the test more readable.